src/main/kotlin/org/rust/ide/typing/paste/RsImportCopyPasteProcessor.kt

```diff
@@ -0,0 +1,213 @@
+        project: Project,
+        editor: Editor,
+        bounds: RangeMarker,
+        caretOffset: Int,
+        indented: Ref<in Boolean>,
+        values: List<RsTextBlockTransferableData>
+    ) {
+        PsiDocumentManager.getInstance(project).commitAllDocuments()
+
+        val data = values.getOrNull(0) ?: return
+        val rsFile = editor.document.toPsiFile(project) as? RsFile ?: return
+        val range = bounds.range
+
+        val elements = gatherElements(rsFile, range)
+        val importCtx = elements.firstOrNull { it is RsElement } as? RsElement ?: return
+
+        val visitor = ImportingVisitor(project, importCtx, range, data)
+
+        runWriteAction {
+            for (element in elements) {
+                element.accept(visitor)
+            }
+        }
+    }
+
+    companion object {
+        val dataFlavor: DataFlavor? by lazy {
+            try {
+                val dataClass = RsReferenceData::class.java
+                DataFlavor(
+                    DataFlavor.javaJVMLocalObjectMimeType + ";class=" + dataClass.name,
+                    "RsReferenceData",
+                    dataClass.classLoader
+                )
+            } catch (e: NoClassDefFoundError) {
+                null
+            } catch (e: IllegalArgumentException) {
+                null
+            }
+        }
+    }
+}
+
+private class RsReferenceData
+
+private class ImportingVisitor(
+    private val project: Project,
+    private val importCtx: RsElement,
+    private val range: TextRange,
+    private val data: RsTextBlockTransferableData
+) : RsRecursiveVisitor() {
+    override fun visitPath(path: RsPath) {
+        val ctx = AutoImportFix.findApplicableContext(project, path)
+        handleContext(path, ctx)
+        super.visitPath(path)
+    }
+
+    override fun visitMethodCall(methodCall: RsMethodCall) {
+        val ctx = AutoImportFix.findApplicableContext(project, methodCall)
+        handleContext(methodCall, ctx)
+        super.visitMethodCall(methodCall)
+    }
+
+    private fun handleContext(element: PsiElement, ctx: AutoImportFix.Context?) {
+        if (ctx != null) {
+            if (ctx.candidates.size == 1) {
+                ctx.candidates[0].import(importCtx)
+            } else {
+                val candidate = ctx.candidates.find {
+                    val offset = toRelativeOffset(element, range)
+                    val fqn = data.offsetToFqnMap[offset]
+                    fqn == it.qualifiedNamedItem.item.qualifiedName
+                }
+                candidate?.import(importCtx)
+            }
+        }
+    }
+}
+
+/**
+ * Records mapping between offsets (relative to copy/paste content range) and fully qualified names of resolved items
+ * from paths and method calls.
+ */
+private fun createFqnMap(file: RsFile, range: TextRange): Map<Int, String> {
+    val elements = gatherElements(file, range)
+    val map = mutableMapOf<Int, String>()
+
+    val visitor = object : RsRecursiveVisitor() {
```

Comment:
I think we also need to handle `RsPatBinding` here, for cases when constants are used in match arms:
```rust
match ... {
    Foo1 => {}
    Foo2 => {}
    _ => {}
}
```

Reply:
`AutoImportFix` doesn't currently supports pat bindings though. Should I try to implement it?

Reply:
I am working on this: #9153